crates/accelerate/src/remove_identity_equiv.rs

```diff
@@ -0,0 +1,113 @@
+// that they have been altered from the originals.
+
+use num_complex::Complex64;
+use num_complex::ComplexFloat;
+use pyo3::prelude::*;
+use rustworkx_core::petgraph::stable_graph::NodeIndex;
+
+use crate::nlayout::PhysicalQubit;
+use crate::target_transpiler::Target;
+use qiskit_circuit::dag_circuit::DAGCircuit;
+use qiskit_circuit::operations::Operation;
+use qiskit_circuit::operations::OperationRef;
+use qiskit_circuit::packed_instruction::PackedInstruction;
+
+#[pyfunction]
+#[pyo3(signature=(dag, approx_degree=Some(1.0), target=None))]
+fn remove_identity_equiv(
+    dag: &mut DAGCircuit,
+    approx_degree: Option<f64>,
+    target: Option<&Target>,
+) {
+    let mut remove_list: Vec<NodeIndex> = Vec::new();
+
+    let get_error_cutoff = |inst: &PackedInstruction| -> f64 {
+        match approx_degree {
+            Some(degree) => {
+                if degree == 1.0 {
+                    f64::EPSILON
+                } else {
+                    match target {
+                        Some(target) => {
+                            let qargs: Vec<PhysicalQubit> = dag
+                                .get_qargs(inst.qubits)
+                                .iter()
+                                .map(|x| PhysicalQubit::new(x.0))
+                                .collect();
+                            let error_rate = target.get_error(inst.op.name(), qargs.as_slice());
+                            match error_rate {
+                                Some(err) => err * degree,
+                                None => degree,
+                            }
+                        }
+                        None => degree,
+                    }
+                }
+            }
+            None => match target {
+                Some(target) => {
+                    let qargs: Vec<PhysicalQubit> = dag
+                        .get_qargs(inst.qubits)
+                        .iter()
+                        .map(|x| PhysicalQubit::new(x.0))
+                        .collect();
+                    let error_rate = target.get_error(inst.op.name(), qargs.as_slice());
+                    match error_rate {
+                        Some(err) => err,
+                        None => f64::EPSILON,
+                    }
+                }
+                None => f64::EPSILON,
+            },
+        }
+    };
+
+    for op_node in dag.op_nodes(false) {
+        let inst = dag.dag()[op_node].unwrap_operation();
+        match inst.op.view() {
+            OperationRef::Standard(gate) => {
+                if let Some(matrix) = gate.matrix(inst.params_view()) {
+                    let error = get_error_cutoff(inst);
+                    let dim = matrix.shape()[0] as f64;
+                    let trace: Complex64 = matrix.diag().iter().sum();
+                    let f_pro = (trace / dim).abs().powi(2);
+                    let gate_fidelity = (dim * f_pro + 1.) / (dim + 1.);
```

Comment:
This would save 1 flop (I don't remember the HPC courses enough though to remember if mult and div have the same cost) 😄 
```suggestion
                    let trace_sq = trace.abs().powi(2);
                    let gate_fidelity = (trace_sq / dim + 1.) / (dim + 1.);
```

Reply:
In hardware in most cases division is more expensive. For example, on my workstation's CPU architecture f64 division is 13 cycles while multiplication is 5 cycles:

https://www.anandtech.com/show/16214/amd-zen-3-ryzen-deep-dive-review-5950x-5900x-5800x-and-5700x-tested/6

but the compiler knows this and will try to optimize away division or use different instructions for the same output if it can so it's not always clear cut. The only way to know for sure is to try it both ways and benchmark it.

Reply:
Compilers often can't optimise away floating-point divisions because of rounding differences. `a / b` isn't the same as `a * (1. / b)` in general (though it is if `b` is a power of two that won't overflow on reciprocation)  - for example, `3.0 / 5.0` isn't the same as `(1.0/5.0) * 3.0` due to rounding - and most transformations suffer similar problems to intermediate rounding.

It's what `-ffast-math` is for in gcc and clang, but we don't run with the equivalent of that in Rust (nor should we).